### PR TITLE
Integration of TicTacToe game into WiFON as separate module. 

### PR DESCRIPTION
**Pull request by spili**
Hi @mikrosk, 
  this PR serves as a "backlog" for all of the things that needs to be done during implementation. Also, you can use it to track my progress as well as review my code if you want to.
<hr>

# What needs to be done:
 - [x] [TicTacToe game logic: Implement the TicTacToe game logic.](https://github.com/croconaut/wifon/pull/4/commits/feab34e4195f64e84e8f84e03055e6bac06a4451) 
 - [x] [TicTacToe game logic tests: Implement and integrate units tests that cover all possible test cases](https://github.com/croconaut/wifon/pull/4/commits/feab34e4195f64e84e8f84e03055e6bac06a4451).
 - [x] Extend current RMB implementation: This covers implementing and integrating new payload messages as well user access to TicTacToe from drawer or conversation with another user.
- [x] UI: implement the UI of the game.

# What would be nice to have:
-  [x] Storage: Create persistent storage for storing the game. Think of additional logic when exactly is this useful. 

<hr> 

  One note ... I'll squash commits into logical units so it is easier for us to track down what has been done already. Also, when this is PR done and therefore ready to merge, I'll rebase and squash all commits into one so we'll keep our repo history clean.

BR,
  Matej

### CODE REVIEW

File: app/src/main/java/com/croconaut/tictactoe/model/Board.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        assertNotNull(listener, "ticTacToeBoardListener");
+        assertIsBiggerThan(size, BOARD_SIZE_MINIMUM, "size"); //$NON-NLS-1$
+
+        mListener = listener;
+        mBoardSize = size;
+        mBoard = new Cell[size][size];
+
+        for (int row = BOARD_INDEX_MINIMUM; row < mBoardSize; row++) {
+            for (int col = BOARD_INDEX_MINIMUM; col < mBoardSize; col++) {
+                mBoard[row][col] = new Cell(row, col, Seed.BLANK);
```
Comment (spili) [inline app/src/main/java/com/croconaut/tictactoe/model/Board.java]: - [ ] use `setCellAt` instead.

File: app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        @Override
+        public void onMovePerformed(boolean movePerformed) {
+            assertThat(movePerformed, is(true));
+        }
+    };
+
+    @NonNull
+    private static final Board.TicTacToeBoardListener LISTENER_EMPTY = new Board.TicTacToeBoardListener() {
```
Comment (spili) [inline app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java]: - [ ] Line Length

File: app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            assertThat(gameState, is(GameState.WIN_NOUGHT));
+        }
+
+        @Override
+        public void onMovePerformed(boolean movePerformed) {
+            assertThat(movePerformed, is(true));
+        }
+    };
+
+    private static final Board.TicTacToeBoardListener LISTENER_WON_CROSS = new Board.TicTacToeBoardListener() {
```
Comment (spili) [inline app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java]: - [ ] Line Length

File: app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+
+
+
+    /* ---------------------------------------- CONSTANTS ----------------------------------------*/
+
+    private static final int FIXTURE_BOARD_SIZE = 3;
+
+    private static final Board.TicTacToeBoardListener LISTENER_WON_NOUGHT = new Board.TicTacToeBoardListener() {
```
Comment (spili) [inline app/src/test/java/com/croconaut/tictactoe/model/BoardTest.java]: - [ ] Line Length

File: app/src/main/java/com/croconaut/tictactoe/model/Board.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+import static com.croconaut.tictactoe.utils.Assertions.assertIsBiggerThan;
+import static com.croconaut.tictactoe.utils.Assertions.assertIsSeed;
+import static com.croconaut.tictactoe.utils.Assertions.assertNotNull;
+
+/**
+ * This class represents TicTacToe game playing field and its logic.
+ * <p>
+ * See also: {@link Cell}, {@link GameState}, {@link Seed}
+ */
+public final class Board {
```
Comment (spili) [inline app/src/main/java/com/croconaut/tictactoe/model/Board.java]: - [ ] Missing reliable and also usable `toString` method. Consider to print using 2D array `mBoard`.

**Comment by spili** [comment]
Also, @mikrosk please [define the mergeability of pull requests](https://help.github.com/articles/defining-the-mergeability-of-pull-requests/), so this won't get merged by accident.

File: app/src/main/java/com/croconaut/tictactoe/model/Board.java (hunk 5 above)
Comment (mikrosk) [reply app/src/main/java/com/croconaut/tictactoe/model/Board.java]: Yep, a mBoardSize x mBoardSize matrix as output seems fine.

**Review by mikrosk — APPROVED** [review]
Looks more professional than anything I've ever written in Java. :-) 

If I understand it correctly, you don't want to merge it yet, right?

**Comment by mikrosk** [comment]
@spili: `master` is now a protected branch.

**Comment by spili** [comment]
> If I understand it correctly, you don't want to merge it yet, right?

No I don't, there's no point in doing so right now. We'll merge when everything regarding the TicTacToe will be done. :-)

> master is now a protected branch.

Great! thanks a lot 👍 

cc. @mikrosk

File: app/src/main/java/com/croconaut/tictactoe/model/Board.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * Creates a new TicTacToe playing field eg. board.
+     *
+     * @param listener Listener for delegating calls eg. creating callback to the caller.
+     * @param size     Requested board size. The minimum board size is {@link #BOARD_INDEX_MINIMUM}.
+     */
+    public Board(@NonNull final TicTacToeBoardListener listener,
+                 @IntRange(from = BOARD_SIZE_MINIMUM) final int size) {
+        assertNotNull(listener, "ticTacToeBoardListener");
```
Comment (spili) [inline app/src/main/java/com/croconaut/tictactoe/model/Board.java]: - [ ] missing `//$NON-NLS-1$`

**Comment by mikrosk** [comment]
Integrated manually.

**Pull request closed without merging**